File: 21b_solution.hs
import qualified Data.Map as M

singleRolls :: [Integer]
singleRolls = [1, 2, 3]

possibleRolls :: [[Integer]]
possibleRolls = do
  a <- singleRolls
  b <- singleRolls
  c <- singleRolls

  return [a, b, c]

type RollValues = M.Map Integer Integer
possibleRollValues :: RollValues

possibleRollValues = foldr reduce M.empty possibleRolls
  where reduce :: [Integer] -> RollValues -> RollValues
        reduce roll values =
          let rollSum = sum roll
              oldVal = M.findWithDefault 0 rollSum values
              newVal = oldVal + 1
          in M.insert rollSum newVal values

move :: Integer -> Integer -> Integer
move a b = ((a-1+b) `mod` 10) + 1

data Player = PlayerA | PlayerB
other :: Player -> Player
other PlayerA = PlayerB
other PlayerB = PlayerA

data PlayerValues = PlayerValues Integer Integer
get :: PlayerValues -> Player -> Integer
get (PlayerValues a _) PlayerA = a
get (PlayerValues _ b) PlayerB = b

set :: PlayerValues -> Player -> Integer -> PlayerValues
set (PlayerValues _ b) PlayerA a = PlayerValues a b
set (PlayerValues a _) PlayerB b = PlayerValues a b

add :: PlayerValues -> PlayerValues -> PlayerValues
add (PlayerValues a1 b1) (PlayerValues a2 b2) = PlayerValues (a1+a2) (b1+b2)

mult :: Integer -> PlayerValues -> PlayerValues
mult c (PlayerValues a b) = PlayerValues (c * a) (c * b)

nulls :: PlayerValues
nulls = PlayerValues 0 0

max' :: PlayerValues -> Integer
max' (PlayerValues a b) = max a b

type Scores = PlayerValues
type Positions = PlayerValues
type Wins = PlayerValues

play :: Player -> Positions -> Scores -> Wins
play player positions scores
  | get scores PlayerA >= 21 = set nulls PlayerA 1
  | get scores PlayerB >= 21 = set nulls PlayerB 1
  | otherwise = M.foldrWithKey reduce nulls possibleRollValues
  where
    reduce :: Integer -> Integer -> PlayerValues -> PlayerValues
    reduce rollVal count wins =
      let playerPos = move (get positions player) rollVal
          newPos = set positions player playerPos
          playerScore = (get scores player) + playerPos
          newScores = set scores player playerScore

          subWins = play (other player) newPos newScores
      in  add wins (mult count subWins)

startPlayer = PlayerA
startPos = PlayerValues 6 3

main :: IO ()
main = do
  let wins = play startPlayer startPos nulls
  putStrLn $ show $ max' wins

File: 21b_solution.rb
start_pos = [6, 3]
start_player = 0

single_rolls = [1,2,3]
possible_rolls = single_rolls.flat_map do |a|
  single_rolls.flat_map do |b|
    single_rolls.map do |c|
      [a, b, c]
    end
  end
end
$possible_roll_values = possible_rolls
  .group_by { |e| e.sum }
  .map { |k, v| [k, v.count] }
  .to_h

def mod_add1(a, b, mod)
  ((a-1+b) % mod) + 1
end

def other(player)
  player == 0 ? 1 : 0
end

def play(player, current_pos, score)
  if score[player] >= 21
    num_wins = []
    num_wins[player] = 1
    num_wins[other(player)] = 0
    return num_wins
  end
  if score[other(player)] >= 21
    num_wins = []
    num_wins[other(player)] = 1
    num_wins[player] = 0
    return num_wins
  end


  $possible_roll_values.reduce([0, 0]) do |num_wins, (value, count)|
    new_pos = []
    new_pos[player] = mod_add1(current_pos[player], value, 10)
    new_pos[other(player)] = current_pos[other(player)]

    new_score = []
    new_score[player] = score[player] + new_pos[player]
    new_score[other(player)] = score[other(player)]

    sub_num_wins = play(other(player), new_pos, new_score)

    [
      num_wins[0]+count*sub_num_wins[0],
      num_wins[1]+count*sub_num_wins[1]
    ]
  end
end

puts play(start_player, start_pos, [0, 0]).max
